x-pack/plugin/index-lifecycle/src/main/java/org/elasticsearch/xpack/indexlifecycle/MoveToNextStepUpdateTask.java

```diff
@@ -52,15 +54,25 @@ String getPolicy() {
     @Override
     public ClusterState execute(ClusterState currentState) {
         Settings indexSettings = currentState.getMetaData().index(index).getSettings();
-        if (policy.equals(LifecycleSettings.LIFECYCLE_NAME_SETTING.get(indexSettings))
-            && currentStepKey.equals(IndexLifecycleRunner.getCurrentStepKey(indexSettings))) {
-            return IndexLifecycleRunner.moveClusterStateToNextStep(index, currentState, currentStepKey, nextStepKey, nowSupplier);
-        } else {
-            // either the policy has changed or the step is now
-            // not the same as when we submitted the update task. In
-            // either case we don't want to do anything now
-            return currentState;
+        String indexPolicySetting = LifecycleSettings.LIFECYCLE_NAME_SETTING.get(indexSettings);
+
+        // policy could be updated in-between execution
+        if (policy.equals(indexPolicySetting) == false) {
+            throw new IllegalArgumentException("policy [" + policy + "] does not match " + LifecycleSettings.LIFECYCLE_NAME
+                + " [" + indexPolicySetting + "]");
+        }
+
+        if (currentStepKey.equals(IndexLifecycleRunner.getCurrentStepKey(indexSettings)) == false) {
+            throw new IllegalArgumentException("index [" + index.getName() + "] is not on current step [" + currentStepKey + "]");
         }
+
+        try {
+            policyStepsRegistry.getStep(policy, nextStepKey);
+        } catch (IllegalStateException e) {
+            throw new IllegalArgumentException(e.getMessage());
+        }
```

Comment:
I'm not sure about throwing exceptions here during the normal execution, I want to think about this a bit more

Reply:
I felt that validating input must be done within the cluster-state thread due to any updates that may have snuck in between the request and the actual operation. This validation needs to bubble back up so we can inform the user of why their operation was a noop. Is this not what the Task's onFailure is meant to be used for? I believe things specific to the cluster-state internals will only be exposed if trace logging is enabled